train.py

```diff
@@ -1,30 +1,100 @@
+            yield tensor_spectro, tensor_label
+
+    train_dataset = tf.data.Dataset.from_generator(generate_subspectrogram, (tf.float32, tf.float32))
+    train_dataset = train_dataset.batch(config.BATCH_SIZE)
+
+    # building the model
+    if config.MODEL == config.ModelEnum.PIANOCKTAIL_GRU:
+        model = PianocktailGRU()
+    elif config.MODEL == config.ModelEnum.BASIC_CNN:
+        model = BasicCNN()
+    model.build(input_shape=(config.BATCH_SIZE, config.SUBSPECTROGRAM_POINTS, config.MEL_BINS))
+    model.summary()
+    optimizer = tf.optimizers.Adam(config.LEARNING_RATE)
+
+    # define metrics
+    train_loss = tf.keras.metrics.Mean(name='train_loss')
+    train_accuracy = tf.keras.metrics.BinaryAccuracy(name='train_accuracy')
+
+    checkpoint, checkpoint_manager = utils.setup_checkpoints(model, optimizer)
+
+    # restore checkpoint
+    if config.RESTORE_CHECKPOINT:
+        checkpoint.restore(checkpoint_manager.latest_checkpoint)
+        print(
+            f"Restored checkpoint. Model {checkpoint.model.name} - epoch {checkpoint.epoch.value()} - iteration {checkpoint.iteration.value()}")
+
     # for test we iterate over samples one by one
     for epoch in range(config.NB_EPOCHS):
 
         print(f"============================ epoch {epoch} =============================")
         checkpoint.epoch.assign(epoch)
 
         for iteration, (spectro, label) in enumerate(train_dataset):
-            predictions, labels = train_step(spectro, label)
+            predictions, labels = train_step(spectro, label, model, optimizer, train_loss, train_accuracy)
 
             # display metrics
             if iteration % 10 == 0:
-                display_and_reset_metrics(train_loss, train_accuracy, predictions, labels, iteration=iteration)
+                utils.display_and_reset_metrics(train_loss, train_accuracy, predictions, labels, iteration=iteration)
 
             # manage checkpoint
             checkpoint.iteration.assign(iteration)
             if iteration % config.SAVE_PERIOD == 0:
                 checkpoint_manager.save()
 
-    save_model(model, epoch)
+    utils.save_model(model, epoch)
+
+
+@tf.function
+def forward_pass(inputs, labels, model):
+    print("tracing forward graph")
+    predictions = model.call(inputs)
+    loss = tf.keras.losses.binary_crossentropy(
+        y_true=labels,
+        y_pred=predictions
+    )
+    return predictions, loss
+
+
+# declaring forward pass and gradient descent
+@tf.function
+def train_step(inputs, labels, model, optimizer, train_loss, train_accuracy):
+    print("tracing train graph")
+    with tf.GradientTape() as tape:
+        predictions_, loss = forward_pass(inputs, labels, model)
+    gradients = tape.gradient(loss, model.trainable_variables)
+    optimizer.apply_gradients(zip(gradients, model.trainable_variables))
+
+    # compute metrics
+    train_loss.update_state(loss)
+    train_accuracy.update_state(labels, predictions_)
+
+    return predictions_, labels
```

Comment:
By the way, what does the `_` at the end of `prediction_` mean here ?

Reply:
I added it because the function used to be defined in the train function where a predictions variable is used. I forgot to change it when the function was separated.